pkg/codegen/go/gen_program.go

```diff
@@ -1166,8 +1166,8 @@ func (g *generator) genResource(w io.Writer, r *pcl.Resource) {
 
 	modOrAlias := g.getModOrAlias(pkg, mod, originalMod)
 
-	instantiate := func(varName, resourceName string, w io.Writer) {
-		if g.scopeTraversalRoots.Has(varName) || strings.HasPrefix(varName, "__") {
+	instantiate := func(varName, blockName, resourceName string, w io.Writer) {
+		if g.scopeTraversalRoots.Has(blockName) || strings.HasPrefix(varName, "__") {
```

Comment:
Maybe a comment worthwhile here -- when is the block name not the resource name/what does this let us achieve to avoid name clashes?